internal/p2p/channel.go

```diff
@@ -214,3 +218,137 @@ func MergedChannelIterator(ctx context.Context, chs ...Channel) *ChannelIterator
 
 	return iter
 }
+
+type libp2pChannelImpl struct {
+	chDesc  *ChannelDescriptor
+	pubsub  *pubsub.PubSub
+	host    host.Host
+	topic   *pubsub.Topic
+	chainID string
+	wrapper Wrapper
+}
+
+func NewLibP2PChannel(chainID string, chDesc *ChannelDescriptor, ps *pubsub.PubSub, h host.Host) (Channel, error) {
+	ch := &libp2pChannelImpl{
+		chDesc:  chDesc,
+		pubsub:  ps,
+		host:    h,
+		chainID: chainID}
+	topic, err := ps.Join(ch.canonicalizedTopicName())
+	if err != nil {
+		return nil, err
+	}
+	ch.topic = topic
+
+	if w, ok := chDesc.MessageType.(Wrapper); ok {
+		ch.wrapper = w
+	}
+
+	// TODO(tychoish) register handlers for
+	// request/response patterns
+
+	return ch, nil
+}
+
+func (ch *libp2pChannelImpl) String() string {
+	return fmt.Sprintf("Channel<%s>", ch.canonicalizedTopicName())
+}
+
+func (ch *libp2pChannelImpl) canonicalizedTopicName() string {
+	return fmt.Sprintf("%s.%s.%d", ch.chainID, ch.chDesc.Name, ch.chDesc.ID)
```

Comment:
Do we need 'name' here?

Reply:
no, but I think it'd be more readable, and I don't think it hurts anything.